{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "                <script type=\"application/javascript\" id=\"jupyter_black\">\n",
       "                (function() {\n",
       "                    if (window.IPython === undefined) {\n",
       "                        return\n",
       "                    }\n",
       "                    var msg = \"WARNING: it looks like you might have loaded \" +\n",
       "                        \"jupyter_black in a non-lab notebook with \" +\n",
       "                        \"`is_lab=True`. Please double check, and if \" +\n",
       "                        \"loading with `%load_ext` please review the README!\"\n",
       "                    console.log(msg)\n",
       "                    alert(msg)\n",
       "                })()\n",
       "                </script>\n",
       "                "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import re\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from protein_name_mapping import dpp_protein_mapping\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dpp4\n",
      "dpp6\n",
      "dpp8\n",
      "dpp9\n",
      "dpp10\n",
      "FAP\n",
      "misc\n"
     ]
    }
   ],
   "source": [
    "base_path = Path(\"../../data/dpp4\")\n",
    "base_path = Path(\"../../data/dpp4\")\n",
    "fasta_path = base_path / \"dpp.fasta\"\n",
    "\n",
    "paths = [\n",
    "    base_path / \"dpp4.json\",\n",
    "    base_path / \"dpp6.json\",\n",
    "    base_path / \"dpp8.json\",\n",
    "    base_path / \"dpp9.json\",\n",
    "    base_path / \"dpp10.json\",\n",
    "    base_path / \"FAP.json\",\n",
    "    base_path / \"misc.json\",\n",
    "]\n",
    "\n",
    "data = dict()\n",
    "for path in paths:\n",
    "    print(path.stem)\n",
    "    with open(path) as file:\n",
    "        load_data = json.load(file)\n",
    "        data[path.stem] = load_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = list()\n",
    "for gene_name, d in data.items():\n",
    "    uids = list()\n",
    "    taxons = list()\n",
    "    markers = list()\n",
    "    protein_names = list()\n",
    "    seqs = list()\n",
    "    genes = list()\n",
    "    annotation_scores = list()\n",
    "    for entry in list(d.values())[0]:\n",
    "        entry_type = re.match(r\".*\\((.*)\\)\", entry[\"entryType\"]).groups()[0]\n",
    "        uid = entry[\"primaryAccession\"]\n",
    "        annotation_score = entry[\"annotationScore\"]\n",
    "        # taxonomy\n",
    "        taxon_id = entry[\"organism\"][\"taxonId\"]\n",
    "        name = entry[\"organism\"][\"scientificName\"]\n",
    "        lineage = entry[\"organism\"][\"lineage\"]\n",
    "        taxonomy = entry[\"organism\"][\"lineage\"] + [name]\n",
    "        seq = entry[\"sequence\"][\"value\"]\n",
    "        if \"X\" in seq:\n",
    "            continue\n",
    "\n",
    "        # existence\n",
    "        protein_existence = entry[\"proteinExistence\"]\n",
    "        if name == \"Homo sapiens\":\n",
    "            marker = name\n",
    "        elif \"Apis\" in taxonomy:\n",
    "            marker = \"Apis\"\n",
    "        elif \"Serpentes\" in taxonomy:\n",
    "            marker = \"Serpentes\"\n",
    "        else:\n",
    "            marker = np.nan\n",
    "\n",
    "        # --- protein name ---\n",
    "        names = entry[\"proteinDescription\"]\n",
    "        if \"recommendedName\" in names:\n",
    "            protein_name = names[\"recommendedName\"][\"fullName\"][\"value\"]\n",
    "        elif \"submissionNames\" in names:\n",
    "            protein_name = names[\"submissionNames\"][0][\"fullName\"][\"value\"]\n",
    "        else:\n",
    "            print(names)\n",
    "            break\n",
    "        # --- marker ---\n",
    "        if \"Venom\" in protein_name:\n",
    "            marker2 = \"venom\"\n",
    "        else:\n",
    "            marker2 = np.nan\n",
    "        uids.append(uid)\n",
    "        taxons.append(taxonomy[-6])\n",
    "        # taxons.append(name)\n",
    "        markers.append(marker)\n",
    "        protein_names.append(protein_name)\n",
    "        seqs.append(seq)\n",
    "        annotation_scores.append(annotation_score)\n",
    "    df = pd.DataFrame(\n",
    "        {\n",
    "            \"uid\": uids,\n",
    "            \"taxon\": taxons,\n",
    "            \"marker\": markers,\n",
    "            \"protein_name\": protein_names,\n",
    "            \"seq\": seqs,\n",
    "            \"annotation_score\": annotation_scores,\n",
    "        }\n",
    "    )\n",
    "    df[\"gene\"] = gene_name\n",
    "    dfs.append(df)\n",
    "\n",
    "df = pd.concat(dfs)\n",
    "df = df.reset_index(drop=True)\n",
    "df = df.drop_duplicates(subset=\"uid\")\n",
    "df[\"protein_name\"] = df[\"protein_name\"].apply(dpp_protein_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: ylabel='Frequency'>"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAc8AAAESCAYAAACb2F7aAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAqFElEQVR4nO3de1SUdf4H8PdwmRHQGQSdGVhBMVEkb6kFs2pnSxZEKi+0my4qGid/6w9clTLjrGlZG142NU2l3WOgp4yN81N3o9QQb5uOqKxXLNLSBoMBfyEzYDHcvr8//PHU5CWeEZgB3q9znnOc5/ud5/l8H4W333kuoxBCCBAREVGLuTm7ACIioo6G4UlERCQTw5OIiEgmhicREZFMDE8iIiKZGJ5EREQyMTyJiIhk8nB2Aa6gqakJpaWl6NGjBxQKhbPLISIiJxBCoLq6GoGBgXBzu/fckuEJoLS0FEFBQc4ug4iIXEBJSQn69Olzzz4MTwA9evQAcOuAqdVqJ1dDRETOYLVaERQUJGXCvTA8AemjWrVazfAkIuriWnL6jhcMERERycTwJCIikonhSUREJBPDk4iISCaGJxERkUwMTyIiIpkYnkRERDIxPImIiGTiQxKIWqjfSx/bvb66Ms5JlRCRs3HmSUREJBPDk4iISCaGJxERkUwMTyIiIpkYnkRERDIxPImIiGRieBIREcnk1PDs168fFArFbUtycjIAoLa2FsnJyfD390f37t0RHx+P8vJyu22YTCbExcXB29sbWq0WixcvRkNDgzOGQ0REXYRTw/PkyZMoKyuTlry8PADA7373OwDAokWL8NFHHyEnJweHDx9GaWkppk6dKr2/sbERcXFxqKurw7Fjx7Bt2zZkZWVh2bJlThkPERF1DQohhHB2Ec0WLlyI3NxcXLp0CVarFb1798aOHTvw9NNPAwC++OILDB48GEajEZGRkdizZw+eeOIJlJaWQqfTAQAyMjKwZMkSXL9+HUql8o77sdlssNls0mur1YqgoCBYLBao1eq2Hyh1SHzCEFHnZrVaodFoWpQFLnPOs66uDu+99x6effZZKBQKFBYWor6+HlFRUVKfsLAwBAcHw2g0AgCMRiOGDh0qBScAxMTEwGq1oqio6K77Sk9Ph0ajkZagoKC2GxgREXU6LhOeu3fvRlVVFWbPng0AMJvNUCqV8PX1teun0+lgNpulPj8Nzub25ra7SUtLg8VikZaSkpLWGwgREXV6LvNg+K1btyI2NhaBgYFtvi+VSgWVStXm+yEios7JJcLzm2++wf79+7Fz505pnV6vR11dHaqqquxmn+Xl5dDr9VKfEydO2G2r+Wrc5j5EcvC8JhG1hEt8bJuZmQmtVou4uB9/UY0aNQqenp7Iz8+X1hUXF8NkMsFgMAAADAYDzp8/j4qKCqlPXl4e1Go1wsPD228ARETUpTh95tnU1ITMzEwkJibCw+PHcjQaDZKSkpCamgo/Pz+o1WrMnz8fBoMBkZGRAIDo6GiEh4dj5syZWL16NcxmM5YuXYrk5GR+LEtERG3G6eG5f/9+mEwmPPvss7e1rVu3Dm5uboiPj4fNZkNMTAw2b94stbu7uyM3Nxfz5s2DwWCAj48PEhMTsWLFivYcAhERdTFOD8/o6Gjc7VbTbt26YdOmTdi0adNd39+3b1988sknbVUeERHRbVzinCcREVFH4vSZJ1FH9dMrc3lVLlHXwpknERGRTAxPIiIimRieREREMjE8iYiIZOIFQ0T38PPH9RERAZx5EhERycbwJCIikonhSUREJBPDk4iISCZeMETUBvj0IaLOjTNPIiIimRieREREMjE8iYiIZGJ4EhERycTwJCIikonhSUREJJPTw/Pbb7/FjBkz4O/vDy8vLwwdOhSnTp2S2oUQWLZsGQICAuDl5YWoqChcunTJbhuVlZVISEiAWq2Gr68vkpKSUFNT095DISKiLsKp4Xnjxg2MGTMGnp6e2LNnDy5evIg333wTPXv2lPqsXr0aGzZsQEZGBgoKCuDj44OYmBjU1tZKfRISElBUVIS8vDzk5ubiyJEjmDt3rjOGREREXYBTH5KwatUqBAUFITMzU1oXEhIi/VkIgfXr12Pp0qWYNGkSAGD79u3Q6XTYvXs3pk2bhs8//xx79+7FyZMnMXr0aADAxo0bMXHiRPz1r39FYGBg+w6KiIg6PafOPP/1r39h9OjR+N3vfgetVouHHnoIf//736X2K1euwGw2IyoqSlqn0WgQEREBo9EIADAajfD19ZWCEwCioqLg5uaGgoKCO+7XZrPBarXaLURERC3l1PD8+uuvsWXLFoSGhmLfvn2YN28e/vSnP2Hbtm0AALPZDADQ6XR279PpdFKb2WyGVqu1a/fw8ICfn5/U5+fS09Oh0WikJSgoqLWHRkREnZhTw7OpqQkjR47EG2+8gYceeghz587Fc889h4yMjDbdb1paGiwWi7SUlJS06f6IiKhzcWp4BgQEIDw83G7d4MGDYTKZAAB6vR4AUF5ebtenvLxcatPr9aioqLBrb2hoQGVlpdTn51QqFdRqtd1CRETUUk4NzzFjxqC4uNhu3Zdffom+ffsCuHXxkF6vR35+vtRutVpRUFAAg8EAADAYDKiqqkJhYaHU58CBA2hqakJEREQ7jIKIiLoap15tu2jRIvz617/GG2+8gd///vc4ceIE/va3v+Fvf/sbAEChUGDhwoV4/fXXERoaipCQELz88ssIDAzE5MmTAdyaqU6YMEH6uLe+vh4pKSmYNm0ar7QlIqI24dTwfPjhh7Fr1y6kpaVhxYoVCAkJwfr165GQkCD1efHFF3Hz5k3MnTsXVVVVGDt2LPbu3Ytu3bpJfd5//32kpKRg/PjxcHNzQ3x8PDZs2OCMIRERURegEEIIZxfhbFarFRqNBhaLhec/u7iffom1HD//wmt+GTZRxyMnC5z+eD4iIqKOhuFJREQkE8OTiIhIJoYnERGRTAxPIiIimRieREREMjE8iYiIZGJ4EhERycTwJCIikonhSUREJBPDk4iISCaGJxERkUwMTyIiIpkYnkRERDIxPImIiGRieBIREcnE8CQiIpKJ4UlERCQTw5OIiEgmp4bnK6+8AoVCYbeEhYVJ7bW1tUhOToa/vz+6d++O+Ph4lJeX223DZDIhLi4O3t7e0Gq1WLx4MRoaGtp7KERE1IV4OLuABx98EPv375dee3j8WNKiRYvw8ccfIycnBxqNBikpKZg6dSqOHj0KAGhsbERcXBz0ej2OHTuGsrIyzJo1C56ennjjjTfafSxERNQ1OD08PTw8oNfrb1tvsViwdetW7NixA48//jgAIDMzE4MHD8bx48cRGRmJTz/9FBcvXsT+/fuh0+kwYsQIvPbaa1iyZAleeeUVKJXK9h4OERF1AU4/53np0iUEBgaif//+SEhIgMlkAgAUFhaivr4eUVFRUt+wsDAEBwfDaDQCAIxGI4YOHQqdTif1iYmJgdVqRVFR0V33abPZYLVa7RYiIqKWcmp4RkREICsrC3v37sWWLVtw5coVjBs3DtXV1TCbzVAqlfD19bV7j06ng9lsBgCYzWa74Gxub267m/T0dGg0GmkJCgpq3YEREVGn5tDHtl9//TX69+9/3zuPjY2V/jxs2DBERESgb9+++PDDD+Hl5XXf27+btLQ0pKamSq+tVisDlIiIWsyhmeeAAQPw2GOP4b333kNtbW2rFePr64uBAwfi8uXL0Ov1qKurQ1VVlV2f8vJy6RypXq+/7erb5td3Oo/aTKVSQa1W2y1EREQt5VB4/uc//8GwYcOQmpoKvV6P//qv/8KJEyfuu5iamhp89dVXCAgIwKhRo+Dp6Yn8/Hypvbi4GCaTCQaDAQBgMBhw/vx5VFRUSH3y8vKgVqsRHh5+3/UQERHdiUPhOWLECLz11lsoLS3Fu+++i7KyMowdOxZDhgzB2rVrcf369RZt54UXXsDhw4dx9epVHDt2DFOmTIG7uzumT58OjUaDpKQkpKam4uDBgygsLMScOXNgMBgQGRkJAIiOjkZ4eDhmzpyJs2fPYt++fVi6dCmSk5OhUqkcGRoREdEvuq8Lhjw8PDB16lTk5ORg1apVuHz5Ml544QUEBQVh1qxZKCsru+f7r127hunTp2PQoEH4/e9/D39/fxw/fhy9e/cGAKxbtw5PPPEE4uPj8eijj0Kv12Pnzp3S+93d3ZGbmwt3d3cYDAbMmDEDs2bNwooVK+5nWERERPekEEIIR9986tQpvPvuu8jOzoaPjw8SExORlJSEa9eu4dVXX4XVam2Vj3PbmtVqhUajgcVi4fnPLq7fSx879L6rK+Puup2ftxGRa5KTBQ5dbbt27VpkZmaiuLgYEydOxPbt2zFx4kS4ud2ayIaEhCArKwv9+vVzZPNEnRqDlajjcyg8t2zZgmeffRazZ89GQEDAHftotVps3br1vooj6gwcnc0SketyKDwvXbr0i32USiUSExMd2TwREZFLc+iCoczMTOTk5Ny2PicnB9u2bbvvooiIiFyZQ+GZnp6OXr163bZeq9Xy20yIiKjTcyg8TSYTQkJCblvft29f6cHuREREnZVD4anVanHu3Lnb1p89exb+/v73XRQREZErcyg8p0+fjj/96U84ePAgGhsb0djYiAMHDmDBggWYNm1aa9dIRETkUhy62va1117D1atXMX78eHh43NpEU1MTZs2axXOeRETU6TkUnkqlEv/4xz/w2muv4ezZs/Dy8sLQoUPRt2/f1q6PiIjI5TgUns0GDhyIgQMHtlYtREREHYJD4dnY2IisrCzk5+ejoqICTU1Ndu0HDhxoleKIiIhckUPhuWDBAmRlZSEuLg5DhgyBQqFo7bqIiIhclkPhmZ2djQ8//BATJ05s7XqIiIhcnkO3qiiVSgwYMKC1ayEiIuoQHArP559/Hm+99Rbu46tAiYiIOiyHPrb97LPPcPDgQezZswcPPvggPD097dp37tzZKsURERG5IofC09fXF1OmTGntWoiIiDoEh8IzMzOztesgIiLqMBw65wkADQ0N2L9/P9555x1UV1cDAEpLS1FTU+PQ9lauXAmFQoGFCxdK62pra5GcnAx/f390794d8fHxKC8vt3ufyWRCXFwcvL29odVqsXjxYjQ0NDg6LCIiol/k0Mzzm2++wYQJE2AymWCz2fDb3/4WPXr0wKpVq2Cz2ZCRkSFreydPnsQ777yDYcOG2a1ftGgRPv74Y+Tk5ECj0SAlJQVTp07F0aNHAdx6WENcXBz0ej2OHTuGsrIyzJo1C56ennzGLhERtRmHZp4LFizA6NGjcePGDXh5eUnrp0yZgvz8fFnbqqmpQUJCAv7+97+jZ8+e0nqLxYKtW7di7dq1ePzxxzFq1ChkZmbi2LFjOH78OADg008/xcWLF/Hee+9hxIgRiI2NxWuvvYZNmzahrq7OkaERERH9IofC89///jeWLl0KpVJpt75fv3749ttvZW0rOTkZcXFxiIqKsltfWFiI+vp6u/VhYWEIDg6G0WgEABiNRgwdOhQ6nU7qExMTA6vViqKiorvu02azwWq12i1EREQt5dDHtk1NTWhsbLxt/bVr19CjR48Wbyc7Oxv/+c9/cPLkydvazGYzlEolfH197dbrdDqYzWapz0+Ds7m9ue1u0tPT8eqrr7a4TiIiop9yaOYZHR2N9evXS68VCgVqamqwfPnyFj+yr6SkBAsWLMD777+Pbt26OVKGw9LS0mCxWKSlpKSkXfdPREQdm0Ph+eabb+Lo0aMIDw9HbW0t/vCHP0gf2a5atapF2ygsLERFRQVGjhwJDw8PeHh44PDhw9iwYQM8PDyg0+lQV1eHqqoqu/eVl5dDr9cDAPR6/W1X3za/bu5zJyqVCmq12m4hIiJqKYc+tu3Tpw/Onj2L7OxsnDt3DjU1NUhKSkJCQoLdBUT3Mn78eJw/f95u3Zw5cxAWFoYlS5YgKCgInp6eyM/PR3x8PACguLgYJpMJBoMBAGAwGPCXv/wFFRUV0Gq1AIC8vDyo1WqEh4c7MjQiIqJf5PCXYXt4eGDGjBkO77hHjx4YMmSI3TofHx/4+/tL65OSkpCamgo/Pz+o1WrMnz8fBoMBkZGRAG59fBweHo6ZM2di9erVMJvNWLp0KZKTk6FSqRyujYiI6F4cCs/t27ffs33WrFkOFfNz69atg5ubG+Lj42Gz2RATE4PNmzdL7e7u7sjNzcW8efNgMBjg4+ODxMRErFixolX2T0REdCcK4cBXo/z0fkwAqK+vx/fffw+lUglvb29UVla2WoHtwWq1QqPRwGKx8PxnF9fvpY8det/VlXEObefn7yMi55GTBQ5dMHTjxg27paamBsXFxRg7diw++OADh4omIiLqKBx+tu3PhYaGYuXKlViwYEFrbZKIiMgltVp4ArcuIiotLW3NTRIREbkchy4Y+te//mX3WgiBsrIyvP322xgzZkyrFEZEROSqHArPyZMn271WKBTo3bs3Hn/8cbz55putURcREZHLcvjZtkRERF2Vww9JIOoMHL01hYi6NofCMzU1tcV9165d68guiIiIXJZD4Xn69GmcPn0a9fX1GDRoEADgyy+/hLu7O0aOHCn1UygUrVMlERGRC3EoPJ988kn06NED27Ztk542dOPGDcyZMwfjxo3D888/36pFEhERuRKHv5IsPT3d7jF9PXv2xOuvv86rbYmIqNNzaOZptVpx/fr129Zfv34d1dXV910UUUfDC4+IuhaHZp5TpkzBnDlzsHPnTly7dg3Xrl3D//zP/yApKQlTp05t7RqJiIhcikMzz4yMDLzwwgv4wx/+gPr6+lsb8vBAUlIS1qxZ06oFEhERuRqHwtPb2xubN2/GmjVr8NVXXwEAHnjgAfj4+LRqcURERK7ovh4MX1ZWhrKyMoSGhsLHxwcOfDUoERFRh+NQeH733XcYP348Bg4ciIkTJ6KsrAwAkJSUxNtUiIio03MoPBctWgRPT0+YTCZ4e3tL65955hns3bu31YojIiJyRQ6d8/z000+xb98+9OnTx259aGgovvnmm1YpjIiIyFU5NPO8efOm3YyzWWVlJVQqVYu3s2XLFgwbNgxqtRpqtRoGgwF79uyR2mtra5GcnAx/f390794d8fHxKC8vt9uGyWRCXFwcvL29odVqsXjxYjQ0NDgyLCIiohZxKDzHjRuH7du3S68VCgWampqwevVqPPbYYy3eTp8+fbBy5UoUFhbi1KlTePzxxzFp0iQUFRUBuPXx8EcffYScnBwcPnwYpaWldveRNjY2Ii4uDnV1dTh27Bi2bduGrKwsLFu2zJFhERERtYhCOHCJ7IULFzB+/HiMHDkSBw4cwFNPPYWioiJUVlbi6NGjeOCBBxwuyM/PD2vWrMHTTz+N3r17Y8eOHXj66acBAF988QUGDx4Mo9GIyMhI7NmzB0888QRKS0uh0+kA3LoHdcmSJbh+/TqUSmWL9mm1WqHRaGCxWKBWqx2unToeZz8Z6OrKOKfun4h+JCcLHJp5DhkyBF9++SXGjh2LSZMm4ebNm5g6dSpOnz7tcHA2NjYiOzsbN2/ehMFgQGFhIerr6xEVFSX1CQsLQ3BwMIxGIwDAaDRi6NChUnACQExMDKxWqzR7vRObzQar1Wq3EBERtZTsC4bq6+sxYcIEZGRk4M9//vN9F3D+/HkYDAbU1taie/fu2LVrF8LDw3HmzBkolUr4+vra9dfpdDCbzQAAs9lsF5zN7c1td5Oeno5XX331vmsnIqKuSfbM09PTE+fOnWu1AgYNGoQzZ86goKAA8+bNQ2JiIi5evNhq27+TtLQ0WCwWaSkpKWnT/RERUefi0Me2M2bMwNatW1ulAKVSiQEDBmDUqFFIT0/H8OHD8dZbb0Gv16Ourg5VVVV2/cvLy6HX6wEAer3+tqtvm18397kTlUolXeHbvBAREbWUQ/d5NjQ04N1338X+/fsxatSo255pu3btWocLampqgs1mw6hRo+Dp6Yn8/HzEx8cDAIqLi2EymWAwGAAABoMBf/nLX1BRUQGtVgsAyMvLg1qtRnh4uMM1EBER3Yus8Pz666/Rr18/XLhwASNHjgQAfPnll3Z9FApFi7eXlpaG2NhYBAcHo7q6Gjt27MChQ4ewb98+aDQaJCUlITU1FX5+flCr1Zg/fz4MBgMiIyMBANHR0QgPD8fMmTOxevVqmM1mLF26FMnJybLuNyUiIpJDVniGhoairKwMBw8eBHDrcXwbNmy47aKdlqqoqMCsWbNQVlYGjUaDYcOGYd++ffjtb38LAFi3bh3c3NwQHx8Pm82GmJgYbN68WXq/u7s7cnNzMW/ePBgMBvj4+CAxMRErVqxwqB4iIqKWkHWfp5ubG8xms/QRqVqtxpkzZ9C/f/82K7A98D7Prov3eRJRsza/z7MZv4KMiIi6IlnhqVAobjunKeccJxERUWcg65ynEAKzZ8+WLsapra3FH//4x9uutt25c2frVUhERORiZIVnYmKi3esZM2a0ajFEREQdgazwzMzMbKs6iIiIOoz7umCIiIioK2J4EhERycTwJCIikonhSUREJBPDk4iISCaGJxERkUwMTyIiIpkYnkRERDIxPImIiGRieBIREcnE8CQiIpKJ4UlERCQTw5OIiEgmhicREZFMTg3P9PR0PPzww+jRowe0Wi0mT56M4uJiuz61tbVITk6Gv78/unfvjvj4eJSXl9v1MZlMiIuLg7e3N7RaLRYvXoyGhob2HAoREXUhTg3Pw4cPIzk5GcePH0deXh7q6+sRHR2NmzdvSn0WLVqEjz76CDk5OTh8+DBKS0sxdepUqb2xsRFxcXGoq6vDsWPHsG3bNmRlZWHZsmXOGBIREXUBCiGEcHYRza5fvw6tVovDhw/j0UcfhcViQe/evbFjxw48/fTTAIAvvvgCgwcPhtFoRGRkJPbs2YMnnngCpaWl0Ol0AICMjAwsWbIE169fh1Kp/MX9Wq1WaDQaWCwWqNXqNh0jOV+/lz52dgmSqyvjnF0CEf0/OVngUuc8LRYLAMDPzw8AUFhYiPr6ekRFRUl9wsLCEBwcDKPRCAAwGo0YOnSoFJwAEBMTA6vViqKiojvux2azwWq12i1EREQt5TLh2dTUhIULF2LMmDEYMmQIAMBsNkOpVMLX19eur06ng9lslvr8NDib25vb7iQ9PR0ajUZagoKCWnk0RETUmblMeCYnJ+PChQvIzs5u832lpaXBYrFIS0lJSZvvk4iIOg8PZxcAACkpKcjNzcWRI0fQp08fab1er0ddXR2qqqrsZp/l5eXQ6/VSnxMnTthtr/lq3OY+P6dSqaBSqVp5FERE1FU4deYphEBKSgp27dqFAwcOICQkxK591KhR8PT0RH5+vrSuuLgYJpMJBoMBAGAwGHD+/HlUVFRIffLy8qBWqxEeHt4+AyEioi7FqTPP5ORk7NixA//85z/Ro0cP6RylRqOBl5cXNBoNkpKSkJqaCj8/P6jVasyfPx8GgwGRkZEAgOjoaISHh2PmzJlYvXo1zGYzli5diuTkZM4uiYioTTg1PLds2QIA+M1vfmO3PjMzE7NnzwYArFu3Dm5uboiPj4fNZkNMTAw2b94s9XV3d0dubi7mzZsHg8EAHx8fJCYmYsWKFe01DCIi6mJc6j5PZ+F9nl0L7/MkojvpsPd5EhERdQQMTyIiIpkYnkRERDK5xH2eRG3Jlc5xElHnwJknERGRTAxPIiIimRieREREMjE8iYiIZGJ4EhERycSrbalT+PkVtXxyDxG1Jc48iYiIZGJ4EhERycSPbalT4oMRiKgtceZJREQkE8OTiIhIJoYnERGRTDznSR0Wz2sSkbNw5klERCQTw5OIiEgmp4bnkSNH8OSTTyIwMBAKhQK7d++2axdCYNmyZQgICICXlxeioqJw6dIluz6VlZVISEiAWq2Gr68vkpKSUFNT046jICKirsap4Xnz5k0MHz4cmzZtumP76tWrsWHDBmRkZKCgoAA+Pj6IiYlBbW2t1CchIQFFRUXIy8tDbm4ujhw5grlz57bXEIiIqAty6gVDsbGxiI2NvWObEALr16/H0qVLMWnSJADA9u3bodPpsHv3bkybNg2ff/459u7di5MnT2L06NEAgI0bN2LixIn461//isDAwDtu22azwWazSa+tVmsrj4yIiDozlz3neeXKFZjNZkRFRUnrNBoNIiIiYDQaAQBGoxG+vr5ScAJAVFQU3NzcUFBQcNdtp6enQ6PRSEtQUFDbDYToHvq99LHdQkQdg8uGp9lsBgDodDq79TqdTmozm83QarV27R4eHvDz85P63ElaWhosFou0lJSUtHL1RETUmXXJ+zxVKhVUKpWzyyAiog7KZWeeer0eAFBeXm63vry8XGrT6/WoqKiwa29oaEBlZaXUh4iIqLW5bHiGhIRAr9cjPz9fWme1WlFQUACDwQAAMBgMqKqqQmFhodTnwIEDaGpqQkRERLvXTEREXYNTP7atqanB5cuXpddXrlzBmTNn4Ofnh+DgYCxcuBCvv/46QkNDERISgpdffhmBgYGYPHkyAGDw4MGYMGECnnvuOWRkZKC+vh4pKSmYNm3aXa+0JSIiul9ODc9Tp07hsccek16npqYCABITE5GVlYUXX3wRN2/exNy5c1FVVYWxY8di79696Natm/Se999/HykpKRg/fjzc3NwQHx+PDRs2tPtY7sfPr7K8ujLOSZUQEVFLKIQQwtlFOJvVaoVGo4HFYoFarW73/TM8HdMZb+3g3z2R88jJApc950lEROSquuStKuS6OAsnoo6AM08iIiKZOPOkDqMznuMkoo6J4UkujYFJRK6I4UlOx4Akoo6G5zyJiIhkYngSERHJxPAkIiKSiec8iVzIT8//8h5XItfFmScREZFMnHkSdRCclRK5DoZnF8VfxEREjuPHtkRERDJx5ukkLX0wgDMelN4as1I+4J2IOjOGpwu6V7B21I9bO2rdzsQnLxG5LoZnByZndufoL+K2mEEyFIioo2N4tpOuEBhdYYxERADDs1Np7/BiWHY8PBdN1Do6TXhu2rQJa9asgdlsxvDhw7Fx40Y88sgjTq2po4RLR6mT7u5e55S78t8v/7NAbaVThOc//vEPpKamIiMjAxEREVi/fj1iYmJQXFwMrVbr7PI6la78i9iVtNbfQ0u305FCh/9GqT0ohBDC2UXcr4iICDz88MN4++23AQBNTU0ICgrC/Pnz8dJLL93W32azwWazSa8tFguCg4NRUlICtVrdanUNWb6v1bZF5EwXXo2xe/3Tf9s/b5PjXttxdB8t/bm7n7qpc7JarQgKCkJVVRU0Gs29O4sOzmazCXd3d7Fr1y679bNmzRJPPfXUHd+zfPlyAYALFy5cuHC5bSkpKfnF7OnwH9v+7//+LxobG6HT6ezW63Q6fPHFF3d8T1paGlJTU6XXTU1NqKyshL+/PxQKRYv33fy/lNaesXYFPHaO47G7Pzx+juvsx04IgerqagQGBv5i3w4fno5QqVRQqVR263x9fR3enlqt7pT/kNoDj53jeOzuD4+f4zrzsfvFj2v/X4d/tm2vXr3g7u6O8vJyu/Xl5eXQ6/VOqoqIiDqzDh+eSqUSo0aNQn5+vrSuqakJ+fn5MBgMTqyMiIg6q07xsW1qaioSExMxevRoPPLII1i/fj1u3ryJOXPmtOl+VSoVli9ffttHwPTLeOwcx2N3f3j8HMdj96NOcasKALz99tvSQxJGjBiBDRs2ICIiwtllERFRJ9RpwpOIiKi9dPhznkRERO2N4UlERCQTw5OIiEgmhicREZFMDE8Hbdq0Cf369UO3bt0QERGBEydOOLskp3vllVegUCjslrCwMKm9trYWycnJ8Pf3R/fu3REfH3/bwy1MJhPi4uLg7e0NrVaLxYsXo6Ghob2H0uaOHDmCJ598EoGBgVAoFNi9e7dduxACy5YtQ0BAALy8vBAVFYVLly7Z9amsrERCQgLUajV8fX2RlJSEmpoauz7nzp3DuHHj0K1bNwQFBWH16tVtPbR28UvHb/bs2bf9W5wwYYJdn656/NLT0/Hwww+jR48e0Gq1mDx5MoqLi+36tNbP6qFDhzBy5EioVCoMGDAAWVlZbT289nN/j2XvmrKzs4VSqRTvvvuuKCoqEs8995zw9fUV5eXlzi7NqZYvXy4efPBBUVZWJi3Xr1+X2v/4xz+KoKAgkZ+fL06dOiUiIyPFr3/9a6m9oaFBDBkyRERFRYnTp0+LTz75RPTq1UukpaU5Yzht6pNPPhF//vOfxc6dOwWA277YYOXKlUKj0Yjdu3eLs2fPiqeeekqEhISIH374QeozYcIEMXz4cHH8+HHx73//WwwYMEBMnz5dardYLEKn04mEhARx4cIF8cEHHwgvLy/xzjvvtNcw28wvHb/ExEQxYcIEu3+LlZWVdn266vGLiYkRmZmZ4sKFC+LMmTNi4sSJIjg4WNTU1Eh9WuNn9euvvxbe3t4iNTVVXLx4UWzcuFG4u7uLvXv3tut42wrD0wGPPPKISE5Oll43NjaKwMBAkZ6e7sSqnG/58uVi+PDhd2yrqqoSnp6eIicnR1r3+eefCwDCaDQKIW79QnRzcxNms1nqs2XLFqFWq4XNZmvT2p3p57/8m5qahF6vF2vWrJHWVVVVCZVKJT744AMhhBAXL14UAMTJkyelPnv27BEKhUJ8++23QgghNm/eLHr27Gl37JYsWSIGDRrUxiNqX3cLz0mTJt31PTx+P6qoqBAAxOHDh4UQrfez+uKLL4oHH3zQbl/PPPOMiImJaeshtQt+bCtTXV0dCgsLERUVJa1zc3NDVFQUjEajEytzDZcuXUJgYCD69++PhIQEmEwmAEBhYSHq6+vtjltYWBiCg4Ol42Y0GjF06FC7b8iJiYmB1WpFUVFR+w7Eia5cuQKz2Wx3rDQaDSIiIuyOla+vL0aPHi31iYqKgpubGwoKCqQ+jz76KJRKpdSn+Uvib9y40U6jcZ5Dhw5Bq9Vi0KBBmDdvHr777jupjcfvRxaLBQDg5+cHoPV+Vo1Go902mvt0lt+TDE+Z7vUVaGaz2UlVuYaIiAhkZWVh79692LJlC65cuYJx48ahuroaZrMZSqXytm+v+elxM5vNdzyuzW1dRfNY7/VvzGw2Q6vV2rV7eHjAz8+PxxPAhAkTsH37duTn52PVqlU4fPgwYmNj0djYCIDHr1lTUxMWLlyIMWPGYMiQIQDQaj+rd+tjtVrxww8/tMVw2lWneLYtuYbY2Fjpz8OGDUNERAT69u2LDz/8EF5eXk6sjLqaadOmSX8eOnQohg0bhgceeACHDh3C+PHjnViZa0lOTsaFCxfw2WefObuUDoczT5n4FWgt5+vri4EDB+Ly5cvQ6/Woq6tDVVWVXZ+fHje9Xn/H49rc1lU0j/Ve/8b0ej0qKirs2hsaGlBZWcnjeQf9+/dHr169cPnyZQA8fgCQkpKC3NxcHDx4EH369JHWt9bP6t36qNXqTvGfaYanTPwKtJarqanBV199hYCAAIwaNQqenp52x624uBgmk0k6bgaDAefPn7f7pZaXlwe1Wo3w8PB2r99ZQkJCoNfr7Y6V1WpFQUGB3bGqqqpCYWGh1OfAgQNoamqSvhDBYDDgyJEjqK+vl/rk5eVh0KBB6NmzZzuNxjVcu3YN3333HQICAgB07eMnhEBKSgp27dqFAwcOICQkxK69tX5WDQaD3Taa+3Sa35POvmKpI8rOzhYqlUpkZWWJixcvirlz5wpfX1+7K8+6oueff14cOnRIXLlyRRw9elRERUWJXr16iYqKCiHErcvfg4ODxYEDB8SpU6eEwWAQBoNBen/z5e/R0dHizJkzYu/evaJ3796d8laV6upqcfr0aXH69GkBQKxdu1acPn1afPPNN0KIW7eq+Pr6in/+85/i3LlzYtKkSXe8VeWhhx4SBQUF4rPPPhOhoaF2t1pUVVUJnU4nZs6cKS5cuCCys7OFt7d3h7/VQoh7H7/q6mrxwgsvCKPRKK5cuSL2798vRo4cKUJDQ0Vtba20ja56/ObNmyc0Go04dOiQ3a0833//vdSnNX5Wm29VWbx4sfj888/Fpk2beKsKCbFx40YRHBwslEqleOSRR8Tx48edXZLTPfPMMyIgIEAolUrxq1/9SjzzzDPi8uXLUvsPP/wg/vu//1v07NlTeHt7iylTpoiysjK7bVy9elXExsYKLy8v0atXL/H888+L+vr69h5Kmzt48KAAcNuSmJgohLh1u8rLL78sdDqdUKlUYvz48aK4uNhuG999952YPn266N69u1Cr1WLOnDmiurrars/Zs2fF2LFjhUqlEr/61a/EypUr22uIbepex+/7778X0dHRonfv3sLT01P07dtXPPfcc7f957arHr87HTcAIjMzU+rTWj+rBw8eFCNGjBBKpVL079/fbh8dHb+SjIiISCae8yQiIpKJ4UlERCQTw5OIiEgmhicREZFMDE8iIiKZGJ5EREQyMTyJiIhkYngSERHJxPAkIiKSieFJREQkE8OTiIhIpv8DnuwA03MS+NMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 500x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df[\"seq\"].str.len().plot(kind=\"hist\", bins=100, figsize=(5, 3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fasta_path, \"w\") as handle:\n",
    "    for idx, row in df.iterrows():\n",
    "        handle.write(f\">{row['uid']}\\n\")\n",
    "        handle.write(f\"{row['seq']}\\n\")\n",
    "df.to_csv(base_path / \"dpp.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seq\n",
       "M    5732\n",
       "L     182\n",
       "S     121\n",
       "I      95\n",
       "V      86\n",
       "F      83\n",
       "Q      62\n",
       "P      61\n",
       "A      57\n",
       "T      52\n",
       "K      52\n",
       "G      49\n",
       "R      44\n",
       "E      42\n",
       "N      42\n",
       "Y      39\n",
       "C      38\n",
       "H      29\n",
       "W      19\n",
       "D      18\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"seq\"].str[0].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mmseqs2: redundancy reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "dpp4     1990\n",
       "dpp9     1617\n",
       "dpp10    1572\n",
       "dpp8      730\n",
       "dpp6      630\n",
       "FAP       330\n",
       "misc       34\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"gene\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = Path(\"../../data/dpp4\")\n",
    "data_name = \"dpp4\"\n",
    "fasta_path = base_path / f\"{data_name}.fasta\"\n",
    "seq_id = 0.9\n",
    "\n",
    "with open(fasta_path, \"w\") as handle:\n",
    "    for idx, row in df.iterrows():\n",
    "        if row[\"gene\"] != data_name:\n",
    "            continue\n",
    "        handle.write(f\">{row['uid']}\\n\")\n",
    "        handle.write(f\"{row['seq']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb ../../data/dpp4/dpp4.fasta ../../data/dpp4/mmseqs/dpp4/DB_dpp4 \n",
      "\n",
      "MMseqs Version:       \t14-7e284\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1920] 0s 51ms\n",
      "Time for merging to DB_dpp4_h: 0h 0m 0s 18ms\n",
      "Time for merging to DB_dpp4: 0h 0m 0s 15ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 190ms\n",
      "Create directory ../../data/dpp4/mmseqs/dpp4/tmp\n",
      "cluster ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/DB_dpp4_clu ../../data/dpp4/mmseqs/dpp4/tmp --min-seq-id 0.9 -c 0.8 --alignment-mode 3 --cov-mode 3 -s 7.5 --comp-bias-corr 0 --mask 0 \n",
      "\n",
      "MMseqs Version:                     \t14-7e284\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t7.5\n",
      "k-mer length                        \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t3\n",
      "Compositional bias                  \t0\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t0\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 3 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 3 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 3 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 1990 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 1.99K 0s 26ms     \n",
      "Sort kmer 0h 0m 0s 3ms\n",
      "Sort by rep. sequence 0h 0m 0s 5ms\n",
      "Time for fill: 0h 0m 0s 1ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 55ms\n",
      "rescorediagonal ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 3 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.99K 0s 8ms      \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 4ms==================>] 99.70% 1.98K eta 0s       \n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "clust ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_rescore1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.99K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 1004 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.99K 0s 6ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.99K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 15ms\n",
      "Total time: 0h 0m 0s 17ms\n",
      "\n",
      "Size of the sequence database: 1990\n",
      "Size of the alignment database: 1990\n",
      "Number of clusters: 1397\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "createsubdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/order_redundancy ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "createsubdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/order_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "filterdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_filter1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_filter2 --filter-file ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 1.40K 0s 6ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 39ms\n",
      "rescorediagonal ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_filter2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 3 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.40K 0s 8ms      \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 101ms\n",
      "align ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pref_rescore2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 3 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1397 type: Aminoacid\n",
      "Target database size: 1397 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 1.40K 0s 265ms    \n",
      "Time for merging to aln: 0h 0m 0s 6ms\n",
      "2276 alignments calculated\n",
      "1692 sequence pairs passed the thresholds (0.743410 of overall calculated)\n",
      "1.211167 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 337ms\n",
      "clust ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/aln ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.40K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 295 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.40K 0s 6ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.40K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 15ms\n",
      "Total time: 0h 0m 0s 16ms\n",
      "\n",
      "Size of the sequence database: 1397\n",
      "Size of the alignment database: 1397\n",
      "Number of clusters: 1151\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "mergeclusters ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/pre_clust ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/linclust/17135940370195632028/clust --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 1.40K 0s 6ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 1.15K 0s 15ms     \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 1.99K 0s 19ms     \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "createsubdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_redundancy ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 3 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1151 type: Aminoacid\n",
      "Estimated memory consumption: 983M\n",
      "Target database size: 1151 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 1.15K 0s 16ms     \n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 1.15K 0s 25ms     \n",
      "Index statistics\n",
      "Entries:          432059\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.006751\n",
      "Top 10 k-mers\n",
      "    NGPWEE\t745\n",
      "    LLHTDN\t705\n",
      "    GTDNFQ\t705\n",
      "    FKGAPV\t653\n",
      "    IWWYYV\t651\n",
      "    ASDRGY\t591\n",
      "    DKRAGW\t488\n",
      "    SGFCAV\t478\n",
      "    YYSYRY\t442\n",
      "    KWYDYT\t413\n",
      "Time for index table init: 0h 0m 0s 782ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1151\n",
      "Target db start 1 to 1151\n",
      "[=================================================================] 100.00% 1.15K 0s 160ms    \n",
      "\n",
      "2.852457 k-mers per position\n",
      "30670 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "19 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 2s 28ms\n",
      "align ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step0 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 3 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1151 type: Aminoacid\n",
      "Target database size: 1151 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 1.15K 2s 900ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 6ms\n",
      "8906 alignments calculated\n",
      "1717 sequence pairs passed the thresholds (0.192791 of overall calculated)\n",
      "1.491746 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 31ms\n",
      "clust ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step0 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.15K 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 552 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.15K 0s 6ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.15K 0s 2ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 23ms\n",
      "Total time: 0h 0m 0s 24ms\n",
      "\n",
      "Size of the sequence database: 1151\n",
      "Size of the alignment database: 1151\n",
      "Number of clusters: 862\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "createsubdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step0 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "prefilter ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 4.25 -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 3 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 60 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 862 type: Aminoacid\n",
      "Estimated memory consumption: 981M\n",
      "Target database size: 862 type: Aminoacid\n",
      "Index table k-mer threshold: 125 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 862 0s 12ms     \n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 862 0s 63ms     \n",
      "Index statistics\n",
      "Entries:          549664\n",
      "DB size:          491 MB\n",
      "Avg k-mer size:   0.008589\n",
      "Top 10 k-mers\n",
      "    NGPWEE\t512\n",
      "    AIGSGY\t474\n",
      "    TADVQQ\t418\n",
      "    IWWYYV\t404\n",
      "    KKPLVY\t390\n",
      "    ASDRGY\t355\n",
      "    IVSDGS\t322\n",
      "    VHQAIS\t307\n",
      "    YLSEIV\t297\n",
      "    DKRAGW\t295\n",
      "Time for index table init: 0h 0m 0s 866ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 125\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 862\n",
      "Target db start 1 to 862\n",
      "[=================================================================] 100.00% 862 1s 288ms    \n",
      "\n",
      "104.162134 k-mers per position\n",
      "55080 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "19 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 2s 991ms\n",
      "align ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 3 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 862 type: Aminoacid\n",
      "Target database size: 862 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 862 1s 998ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 6ms\n",
      "5718 alignments calculated\n",
      "866 sequence pairs passed the thresholds (0.151452 of overall calculated)\n",
      "1.004640 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 26ms\n",
      "clust ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 862 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 4 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 862 0s 4ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 862 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 11ms\n",
      "Total time: 0h 0m 0s 12ms\n",
      "\n",
      "Size of the sequence database: 862\n",
      "Size of the alignment database: 862\n",
      "Number of clusters: 858\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "createsubdb ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "prefilter ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 7.5 -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 3 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 60 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 858 type: Aminoacid\n",
      "Estimated memory consumption: 981M\n",
      "Target database size: 858 type: Aminoacid\n",
      "Index table k-mer threshold: 96 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 858 0s 17ms     \n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 858 0s 31ms     \n",
      "Index statistics\n",
      "Entries:          549791\n",
      "DB size:          491 MB\n",
      "Avg k-mer size:   0.008590\n",
      "Top 10 k-mers\n",
      "    NGPWEE\t509\n",
      "    AIGSGY\t471\n",
      "    TADVQQ\t415\n",
      "    IWWYYV\t401\n",
      "    KKPLVY\t387\n",
      "    ASDRGY\t353\n",
      "    IVSDGS\t320\n",
      "    VHQAIS\t304\n",
      "    YLSEIV\t295\n",
      "    DKRAGW\t292\n",
      "Time for index table init: 0h 0m 0s 671ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 96\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 858\n",
      "Target db start 1 to 858\n",
      "[=================================================================] 100.00% 858 17s 56ms     \n",
      "\n",
      "2108.059471 k-mers per position\n",
      "112038 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "19 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 18s 847ms\n",
      "align ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/pref_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 3 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 858 type: Aminoacid\n",
      "Target database size: 858 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 858 2s 208ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 6ms\n",
      "5699 alignments calculated\n",
      "858 sequence pairs passed the thresholds (0.150553 of overall calculated)\n",
      "1.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 244ms\n",
      "clust ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/input_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/aln_step2 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 858 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 0 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 858 0s 2ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 858 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 10ms\n",
      "Total time: 0h 0m 0s 11ms\n",
      "\n",
      "Size of the sequence database: 858\n",
      "Size of the alignment database: 858\n",
      "Number of clusters: 858\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "mergeclusters ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/DB_dpp4_clu ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_redundancy ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step0 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step1 ../../data/dpp4/mmseqs/dpp4/tmp/16436520784258545986/clu_step2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 1.15K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 862 0s 9ms      \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 858 0s 14ms     \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 858 0s 18ms     \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 1.99K 0s 24ms     \n",
      "Time for merging to DB_dpp4_clu: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "createtsv ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/DB_dpp4 ../../data/dpp4/mmseqs/dpp4/DB_dpp4_clu ../../data/dpp4/mmseqs/dpp4/DB_dpp4_clu.tsv \n",
      "\n",
      "MMseqs Version:                 \t14-7e284\n",
      "First sequence as representative\tfalse\n",
      "Target column                   \t1\n",
      "Add full header                 \tfalse\n",
      "Sequence source                 \t0\n",
      "Database output                 \tfalse\n",
      "Threads                         \t8\n",
      "Compressed                      \t0\n",
      "Verbosity                       \t3\n",
      "\n",
      "Time for merging to DB_dpp4_clu.tsv: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 12ms\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "\n",
    "mmseqs_dir = base_path / \"mmseqs\"\n",
    "work_dir = mmseqs_dir / data_name\n",
    "db_path = work_dir / f\"DB_{data_name}\"\n",
    "clu_path = work_dir / f\"DB_{data_name}_clu\"\n",
    "tsv_path = work_dir / f\"DB_{data_name}_clu.tsv\"\n",
    "tmp_path = work_dir / \"tmp\"\n",
    "\n",
    "if work_dir.exists():\n",
    "    shutil.rmtree(work_dir)\n",
    "work_dir.mkdir(exist_ok=True)\n",
    "# !rm -r {work_dir}\n",
    "# !mkdir {work_dir}\n",
    "\n",
    "!mmseqs createdb {fasta_path} {db_path}\n",
    "!mmseqs cluster {db_path} {clu_path} {tmp_path} --min-seq-id {seq_id} -c 0.8 --alignment-mode 3 --cov-mode 3 -s 7.5 --comp-bias-corr 0 --mask 0\n",
    "!mmseqs createtsv {db_path} {db_path} {clu_path} {tsv_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_clusters(cluster_tsv, df):\n",
    "    df_clu = pd.read_csv(\n",
    "        cluster_tsv, sep=\"\\t\", header=None, names=[\"representative\", \"member\"]\n",
    "    )\n",
    "    rep = df_clu[\"representative\"].unique()\n",
    "    df_rep = df.loc[df[\"uid\"].isin(rep), :]\n",
    "    # df_rep[\"Pharmacological Family\"].value_counts()\n",
    "    return df_clu, df_rep\n",
    "\n",
    "\n",
    "# df = pd.DataFrame(list(seqs.items()), columns=[\"uid\", \"seq\"])\n",
    "df_clu, df_rep = parse_clusters(cluster_tsv=tsv_path, df=df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x1419e0b10>"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clu.groupby(\"representative\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "# Function to run shell commands\n",
    "def run_command(command):\n",
    "    process = subprocess.Popen(\n",
    "        command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "    )\n",
    "    stdout, stderr = process.communicate()\n",
    "    return stdout, stderr\n",
    "\n",
    "\n",
    "# Step 1: Cluster sequences with MMseqs2\n",
    "mmseqs_command = \"mmseqs easy-cluster input.fasta clusterResult tmp --min-seq-id 0.9\"\n",
    "stdout, stderr = run_command(mmseqs_command)\n",
    "if stderr:\n",
    "    print(\"Error in MMseqs2 clustering:\", stderr.decode())\n",
    "else:\n",
    "    print(\"Clustering completed.\")\n",
    "\n",
    "# Step 2: Read the clustering result\n",
    "cluster_file = \"clusterResult_cluster.tsv\"\n",
    "if not os.path.exists(cluster_file):\n",
    "    print(\"Cluster file not found.\")\n",
    "else:\n",
    "    # Read cluster file\n",
    "    clusters = pd.read_csv(\n",
    "        cluster_file, sep=\"\\t\", header=None, names=[\"cluster_id\", \"sequence_id\"]\n",
    "    )\n",
    "\n",
    "    # Count sequences in each cluster\n",
    "    cluster_counts = clusters[\"cluster_id\"].value_counts()\n",
    "\n",
    "    # Identify singletons and small clusters\n",
    "    outlier_clusters = cluster_counts[\n",
    "        cluster_counts <= 3\n",
    "    ].index.tolist()  # Adjust the threshold as needed\n",
    "\n",
    "    # Filter out outliers\n",
    "    outliers = clusters[clusters[\"cluster_id\"].isin(outlier_clusters)]\n",
    "\n",
    "    # Step 3: Remove outliers from original dataset\n",
    "    def filter_sequences(input_file, output_file, exclude_ids):\n",
    "        sequences = SeqIO.parse(input_file, \"fasta\")\n",
    "        filtered = (seq for seq in sequences if seq.id not in exclude_ids)\n",
    "        SeqIO.write(filtered, output_file, \"fasta\")\n",
    "\n",
    "    filter_sequences(\"input.fasta\", \"cleaned.fasta\", set(outliers[\"sequence_id\"]))\n",
    "    print(\"Outlier sequences removed. Cleaned dataset saved as 'cleaned.fasta'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_clusters(cluster_tsv, df):\n",
    "    df_clu = pd.read_csv(\n",
    "        cluster_tsv, sep=\"\\t\", header=None, names=[\"representative\", \"member\"]\n",
    "    )\n",
    "    rep = df_clu[\"representative\"].unique()\n",
    "    df_rep = df.loc[df[\"uid\"].isin(rep), :]\n",
    "    # df_rep[\"Pharmacological Family\"].value_counts()\n",
    "    return df_clu, df_rep\n",
    "\n",
    "\n",
    "df = pd.DataFrame(list(seqs.items()), columns=[\"uid\", \"seq\"])\n",
    "df_clu, df_rep = parse_clusters(cluster_tsv=output_paths[\"createtsv\"], df=df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(base_path / \"dpp4_90.fasta\", \"w\") as handle:\n",
    "    for idx, row in df_rep.iterrows():\n",
    "        handle.write(f\">{row['uid']}\\n\")\n",
    "        handle.write(f\"{row['seq']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "307"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_rep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "uids = list()\n",
    "taxons = list()\n",
    "for entry in list(dpp4_data.values())[0]:\n",
    "    entry_type = re.match(r\".*\\((.*)\\)\", entry[\"entryType\"]).groups()[0]\n",
    "    uid = entry[\"primaryAccession\"]\n",
    "    annotation_score = entry[\"annotationScore\"]\n",
    "    uids.append(uid)\n",
    "    # taxonomy\n",
    "    taxon_id = entry[\"organism\"][\"taxonId\"]\n",
    "    name = entry[\"organism\"][\"scientificName\"]\n",
    "    lineage = entry[\"organism\"][\"lineage\"]\n",
    "    taxonomy = entry[\"organism\"][\"lineage\"] + [entry[\"organism\"][\"scientificName\"]]\n",
    "    taxons.append(taxonomy[4])\n",
    "    # taxons.append(name)\n",
    "    # existence\n",
    "    protein_existence = entry[\"proteinExistence\"]\n",
    "    # if \"keywords\" in entry:\n",
    "    #     print(entry[\"keywords\"])\n",
    "    #     break\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KW-0472 Membrane Cellular component\n",
      "KW-0812 Transmembrane Domain\n",
      "KW-1133 Transmembrane helix Domain\n"
     ]
    }
   ],
   "source": [
    "for kw in entry[\"keywords\"]:\n",
    "    name = kw[\"name\"]\n",
    "    category = kw[\"category\"]\n",
    "    kw_id = kw[\"id\"]\n",
    "    print(kw_id, name, category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({\"uids\": uids, \"taxon\": taxons})\n",
    "df.to_csv(base_path / \"dpp4_tmp.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Vertebrata': 1784,\n",
       " 'Hexapoda': 522,\n",
       " 'Mollusca': 59,\n",
       " 'Chelicerata': 54,\n",
       " 'Crustacea': 27,\n",
       " 'Chromadorea': 14,\n",
       " 'Leptocardii': 10,\n",
       " 'Platyhelminthes': 9,\n",
       " 'Hexacorallia': 6,\n",
       " 'Enoplea': 5,\n",
       " 'Hydroidolina': 3,\n",
       " 'Myxosporea': 2,\n",
       " 'Trichoplacea': 2,\n",
       " 'Octocorallia': 2,\n",
       " 'Eutardigrada': 1,\n",
       " 'Echinozoa': 1}"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"taxon\"].value_counts().to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- get the 10 species with the largest taxonomic distance between them having a BUSCO score of at least 90% from all Metazoas\n",
    "- Distance defenition:\n",
    "  - travers the taxonomy tree where the distance to each parent and to each child is 1\n",
    "    - vectorize: `set(taxon1) ^ set(taxon2)`\n",
    "  - get years it took to evolve from one species to another"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['entryType', 'primaryAccession', 'uniProtkbId', 'entryAudit', 'annotationScore', 'organism', 'proteinExistence', 'proteinDescription', 'genes', 'comments', 'features', 'keywords', 'references', 'uniProtKBCrossReferences', 'sequence', 'extraAttributes'])"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entry.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_taxon_distance(taxon1, taxon2):\n",
    "    return len(set(taxon1) ^ set(taxon2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb Cell 7\u001b[0m line \u001b[0;36m9\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m arr \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mzeros((\u001b[39mlen\u001b[39m(taxons), \u001b[39mlen\u001b[39m(taxons)))\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39mfor\u001b[39;00m (idx, jdx), (taxon1, taxon2) \u001b[39min\u001b[39;00m \u001b[39mzip\u001b[39m(\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m     itertools\u001b[39m.\u001b[39mcombinations(\u001b[39mrange\u001b[39m(\u001b[39mlen\u001b[39m(taxons)), \u001b[39m2\u001b[39m),\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m     itertools\u001b[39m.\u001b[39mcombinations(taxons, \u001b[39m2\u001b[39m),\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m ):\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     arr[idx, jdx] \u001b[39m=\u001b[39m calc_taxon_distance(taxon1, taxon2)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_dpp4.ipynb#X15sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m arr\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "arr = np.zeros((len(taxons), len(taxons)))\n",
    "for (idx, jdx), (taxon1, taxon2) in zip(\n",
    "    itertools.combinations(range(len(taxons)), 2),\n",
    "    itertools.combinations(taxons, 2),\n",
    "):\n",
    "    arr[idx, jdx] = calc_taxon_distance(taxon1, taxon2)\n",
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(taxonomy))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create a graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "\n",
    "def build_graph_from_taxons(taxons):\n",
    "    G = nx.DiGraph()\n",
    "    for lineage in taxons:\n",
    "        for i in range(len(lineage) - 1):\n",
    "            parent, child = lineage[i], lineage[i + 1]\n",
    "            G.add_edge(parent, child)\n",
    "    return G\n",
    "\n",
    "\n",
    "def find_leaf_nodes(G):\n",
    "    return [node for node, degree in G.out_degree() if degree == 0]\n",
    "\n",
    "\n",
    "def most_distant_leafs(G, x):\n",
    "    all_paths = dict(nx.floyd_warshall(G))\n",
    "    leaf_nodes = find_leaf_nodes(G)\n",
    "    distant_leafs = []\n",
    "\n",
    "    for i in range(len(leaf_nodes)):\n",
    "        for j in range(i + 1, len(leaf_nodes)):\n",
    "            distance = all_paths[leaf_nodes[i]][leaf_nodes[j]]\n",
    "            distant_leafs.append((leaf_nodes[i], leaf_nodes[j], distance))\n",
    "\n",
    "    distant_leafs.sort(key=lambda x: x[2], reverse=True)\n",
    "    return distant_leafs[:x]\n",
    "\n",
    "\n",
    "# Your taxons data\n",
    "# taxons = [\n",
    "#     ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Artiodactyla', 'Ruminantia', 'Pecora', 'Bovidae', 'Caprinae', 'Capra', 'Capra hircus'],\n",
    "#     ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Feliformia', 'Felidae', 'Felinae', 'Felis', 'Felis catus'],\n",
    "#     # Add the rest of your taxons here\n",
    "# ]\n",
    "\n",
    "G = build_graph_from_taxons(taxons)\n",
    "# x_most_distant = most_distant_leafs(G, 2)\n",
    "# print(x_most_distant)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy 2:\n",
    "- **Pre-compute Distance to Leaf:** While building the graph, you can also calculate and store the distance of each node from the leaf.\n",
    "This can be done by a simple depth-first traversal from the root.\n",
    "- **Modify Graph Nodes:** Each node in the graph can be a tuple (species, distance_to_leaf).\n",
    "When searching for distant leaves, you can prioritize paths that lead to nodes with greater distance_to_leaf."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most distant leaves: ['Capra hircus', 'Felis catus', 'Canis lupus familiaris', 'Bos taurus', 'Monodelphis domestica', 'Mus musculus', 'Rattus norvegicus', 'Sus scrofa', 'Homo sapiens', 'Glycine max']\n"
     ]
    }
   ],
   "source": [
    "def build_graph_and_compute_distances(taxons):\n",
    "    G = nx.DiGraph()\n",
    "    max_depth = {\n",
    "        node: 0 for lineage in taxons for node in lineage\n",
    "    }  # Initialize max depth\n",
    "\n",
    "    for lineage in taxons:\n",
    "        for i in range(len(lineage) - 1):\n",
    "            parent, child = lineage[i], lineage[i + 1]\n",
    "            G.add_edge(parent, child)\n",
    "            max_depth[child] = max(max_depth[child], len(lineage) - i - 1)\n",
    "\n",
    "    return G, max_depth\n",
    "\n",
    "\n",
    "def find_most_distant_leaves(G, max_depth, x):\n",
    "    leaves = [node for node in G.nodes if G.out_degree(node) == 0]\n",
    "    leaves.sort(key=lambda leaf: max_depth[leaf], reverse=True)\n",
    "\n",
    "    # Assuming the tree is balanced, the most distant leaves should be among those with the greatest depth\n",
    "    return leaves[:x]\n",
    "\n",
    "\n",
    "G, max_depth = build_graph_and_compute_distances(taxons)\n",
    "most_distant_leaves = find_most_distant_leaves(\n",
    "    G, max_depth, 10\n",
    ")  # Change 2 to the number of leaves you want\n",
    "print(\"Most distant leaves:\", most_distant_leaves)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy 1:\n",
    "- Identify Divergence Points: You can identify the earliest divergence points in the tree where the paths to different leaves start to diverge.\n",
    "- Traverse Downward from These Points: For each divergence point, traverse down the two most distinct paths. The leaves at the ends of these paths are likely to be among the most distant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most distant leaves: ('Calcarius ornatus', 'Glycine max')\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def build_graph(taxons):\n",
    "    G = nx.DiGraph()\n",
    "    for lineage in taxons:\n",
    "        for i in range(len(lineage) - 1):\n",
    "            parent, child = lineage[i], lineage[i + 1]\n",
    "            G.add_edge(parent, child)\n",
    "    return G\n",
    "\n",
    "\n",
    "def find_divergence_points(G):\n",
    "    divergence_points = []\n",
    "    for node in G.nodes:\n",
    "        if G.out_degree(node) > 1:  # More than one child signifies a divergence point\n",
    "            divergence_points.append(node)\n",
    "    return divergence_points\n",
    "\n",
    "\n",
    "def find_distant_leaves_from_divergence(G, divergence_node):\n",
    "    paths = nx.single_source_shortest_path(G, divergence_node)\n",
    "    leaf_paths = {end: path for end, path in paths.items() if G.out_degree(end) == 0}\n",
    "\n",
    "    if not leaf_paths:\n",
    "        return None  # No leaves found from this divergence point\n",
    "\n",
    "    # Find two leaves that are most distant from each other\n",
    "    max_distance = 0\n",
    "    most_distant_pair = None\n",
    "    for leaf1 in leaf_paths:\n",
    "        for leaf2 in leaf_paths:\n",
    "            if leaf1 != leaf2:\n",
    "                distance = len(\n",
    "                    set(leaf_paths[leaf1]) ^ set(leaf_paths[leaf2])\n",
    "                )  # Unique nodes in both paths\n",
    "                if distance > max_distance:\n",
    "                    max_distance = distance\n",
    "                    most_distant_pair = (leaf1, leaf2)\n",
    "    return most_distant_pair\n",
    "\n",
    "\n",
    "# Your taxons data\n",
    "G = build_graph(taxons)\n",
    "divergence_points = find_divergence_points(G)\n",
    "most_distant_leaves = []\n",
    "\n",
    "for divergence_point in divergence_points:\n",
    "    leaves = find_distant_leaves_from_divergence(G, divergence_point)\n",
    "    if leaves:\n",
    "        most_distant_leaves.append(leaves)\n",
    "\n",
    "if most_distant_leaves:\n",
    "    most_distant_leaves = sorted(\n",
    "        most_distant_leaves,\n",
    "        key=lambda pair: len(set(G.nodes[pair[0]]) ^ set(G.nodes[pair[1]])),\n",
    "        reverse=True,\n",
    "    )[0]\n",
    "    print(\"Most distant leaves:\", most_distant_leaves)\n",
    "else:\n",
    "    print(\"No distant leaves found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### strategy 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group Representatives: {'Eukaryota': 'Homo sapiens', 'Chordata': 'Homo sapiens'}\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "\n",
    "def build_graph(taxons):\n",
    "    G = nx.DiGraph()\n",
    "    for lineage in taxons:\n",
    "        for i in range(len(lineage) - 1):\n",
    "            parent, child = lineage[i], lineage[i + 1]\n",
    "            G.add_edge(parent, child)\n",
    "    return G\n",
    "\n",
    "\n",
    "def find_divergence_points(G, x):\n",
    "    divergence_points = []\n",
    "    for node in G.nodes:\n",
    "        if G.out_degree(node) > 1:\n",
    "            divergence_points.append(node)\n",
    "            if len(divergence_points) == x:\n",
    "                break\n",
    "    return divergence_points\n",
    "\n",
    "\n",
    "def group_leaves_by_divergence(G, divergence_points):\n",
    "    groups = {}\n",
    "    for point in divergence_points:\n",
    "        leaves = [\n",
    "            n\n",
    "            for n, d in G.nodes(data=True)\n",
    "            if G.out_degree(n) == 0 and nx.has_path(G, point, n)\n",
    "        ]\n",
    "        groups[point] = leaves\n",
    "    return groups\n",
    "\n",
    "\n",
    "def select_representatives(groups, leaf_data):\n",
    "    representatives = {}\n",
    "    for group, leaves in groups.items():\n",
    "        best_leaf = None\n",
    "        best_score = -1\n",
    "        best_cdp = float(\"inf\")\n",
    "        least_missing = float(\"inf\")\n",
    "\n",
    "        for leaf in leaves:\n",
    "            data = leaf_data.get(leaf, {})\n",
    "            busco_score = data.get(\"BUSCO\", -1)\n",
    "            cdp_status = data.get(\"CDP status\", \"Unknown\")\n",
    "            missing = data.get(\"missing\", float(\"inf\"))\n",
    "\n",
    "            cdp_order = [\n",
    "                \"Standard\",\n",
    "                \"Close to standard (high value)\",\n",
    "                \"Close to standard (low value)\",\n",
    "                \"Outlier (high value)\",\n",
    "                \"Outlier (low value)\",\n",
    "                \"Unknown\",\n",
    "            ]\n",
    "            cdp_rank = cdp_order.index(cdp_status)\n",
    "\n",
    "            if (\n",
    "                (busco_score > best_score)\n",
    "                or (busco_score == best_score and cdp_rank < best_cdp)\n",
    "                or (\n",
    "                    busco_score == best_score\n",
    "                    and cdp_rank == best_cdp\n",
    "                    and missing < least_missing\n",
    "                )\n",
    "            ):\n",
    "                best_leaf = leaf\n",
    "                best_score = busco_score\n",
    "                best_cdp = cdp_rank\n",
    "                least_missing = missing\n",
    "\n",
    "        representatives[group] = best_leaf if best_leaf else random.choice(leaves)\n",
    "    return representatives\n",
    "\n",
    "\n",
    "# Example Usage\n",
    "taxons = [\n",
    "    [\n",
    "        \"Eukaryota\",\n",
    "        \"Animalia\",\n",
    "        \"Chordata\",\n",
    "        \"Mammalia\",\n",
    "        \"Primates\",\n",
    "        \"Hominidae\",\n",
    "        \"Homo\",\n",
    "        \"Homo sapiens\",\n",
    "    ],\n",
    "    [\n",
    "        \"Eukaryota\",\n",
    "        \"Animalia\",\n",
    "        \"Chordata\",\n",
    "        \"Mammalia\",\n",
    "        \"Carnivora\",\n",
    "        \"Felidae\",\n",
    "        \"Felis\",\n",
    "        \"Felis catus\",\n",
    "    ],\n",
    "    [\n",
    "        \"Eukaryota\",\n",
    "        \"Animalia\",\n",
    "        \"Chordata\",\n",
    "        \"Aves\",\n",
    "        \"Psittaciformes\",\n",
    "        \"Psittacidae\",\n",
    "        \"Ara\",\n",
    "        \"Ara macao\",\n",
    "    ],\n",
    "    [\n",
    "        \"Eukaryota\",\n",
    "        \"Animalia\",\n",
    "        \"Chordata\",\n",
    "        \"Mammalia\",\n",
    "        \"Carnivora\",\n",
    "        \"Canidae\",\n",
    "        \"Canis\",\n",
    "        \"Canis lupus familiaris\",\n",
    "    ],\n",
    "    [\n",
    "        \"Eukaryota\",\n",
    "        \"Plantae\",\n",
    "        \"Magnoliophyta\",\n",
    "        \"Rosales\",\n",
    "        \"Rosaceae\",\n",
    "        \"Rosa\",\n",
    "        \"Rosa chinensis\",\n",
    "    ],\n",
    "]\n",
    "\n",
    "\n",
    "leaf_data = {\n",
    "    \"Homo sapiens\": {\"BUSCO\": 95, \"CDP status\": \"Standard\", \"missing\": 0},\n",
    "    \"Felis catus\": {\"BUSCO\": 92, \"CDP status\": \"Outlier (high value)\", \"missing\": 5},\n",
    "    \"Ara macao\": {\n",
    "        \"BUSCO\": 90,\n",
    "        \"CDP status\": \"Close to standard (low value)\",\n",
    "        \"missing\": 2,\n",
    "    },\n",
    "    \"Canis lupus familiaris\": {\"BUSCO\": 95, \"CDP status\": \"Standard\", \"missing\": 1},\n",
    "    \"Rosa chinensis\": {\"BUSCO\": 88, \"CDP status\": \"Unknown\", \"missing\": 10},\n",
    "}\n",
    "\n",
    "G = build_graph(taxons)\n",
    "divergence_points = find_divergence_points(\n",
    "    G, 2\n",
    ")  # Replace 'x' with the number of groups you want\n",
    "groups = group_leaves_by_divergence(G, divergence_points)\n",
    "representatives = select_representatives(groups, leaf_data)\n",
    "print(\"Group Representatives:\", representatives)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rostlab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
