{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "                <script type=\"application/javascript\" id=\"jupyter_black\">\n",
       "                (function() {\n",
       "                    if (window.IPython === undefined) {\n",
       "                        return\n",
       "                    }\n",
       "                    var msg = \"WARNING: it looks like you might have loaded \" +\n",
       "                        \"jupyter_black in a non-lab notebook with \" +\n",
       "                        \"`is_lab=True`. Please double check, and if \" +\n",
       "                        \"loading with `%load_ext` please review the README!\"\n",
       "                    console.log(msg)\n",
       "                    alert(msg)\n",
       "                })()\n",
       "                </script>\n",
       "                "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# proteom_json_path = \"../../data/proteomes_proteome_type_1_AND_busco_90_2023_11_09.json\"\n",
    "proteom_json_path = \"../../data/proteomes_proteome_type_1_AND_busco_90_AND_taxonomy_id_332082023_11_09.json\"\n",
    "\n",
    "with open(proteom_json_path) as file:\n",
    "    proteom_data = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxon ID `6279` has no Genome ID\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "\n",
    "for entry in list(proteom_data.values())[0]:\n",
    "    proteome_id = entry[\"id\"]\n",
    "    # proteam score\n",
    "    proteom_score = entry[\"proteomeCompletenessReport\"]\n",
    "    busco = proteom_score[\"buscoReport\"][\"score\"]\n",
    "    busco_nr_proteins = proteom_score[\"buscoReport\"][\"total\"]\n",
    "    cpdReport = proteom_score[\"cpdReport\"][\"status\"]\n",
    "    # taxonomy\n",
    "    taxonomy = entry[\"taxonomy\"]\n",
    "    name = taxonomy[\"scientificName\"]\n",
    "    taxon_id = taxonomy[\"taxonId\"]\n",
    "    # genome\n",
    "    genome = entry[\"genomeAssembly\"]\n",
    "    if \"assemblyId\" not in genome:\n",
    "        print(f\"Taxon ID `{taxon_id}` has no Genome ID\")\n",
    "        continue\n",
    "    genome_id = genome[\"assemblyId\"]\n",
    "\n",
    "    data.append(\n",
    "        (\n",
    "            proteome_id,\n",
    "            taxon_id,\n",
    "            name,\n",
    "            busco,\n",
    "            # busco_nr_proteins,\n",
    "            cpdReport,\n",
    "            genome_id,\n",
    "        )\n",
    "    )\n",
    "    # TODO: display for debugging\n",
    "    # print(name, \" -> \", busco)\n",
    "    # print(genome_id)\n",
    "    # break\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    data,\n",
    "    columns=[\n",
    "        \"proteome_id\",\n",
    "        \"taxon_id\",\n",
    "        \"name\",\n",
    "        \"busco\",\n",
    "        # \"busco_nr_proteins\",\n",
    "        \"cpdReport\",\n",
    "        \"genome_id\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "# make `cpdReport` categorical\n",
    "categories = [\n",
    "    \"Standard\",\n",
    "    \"Close to standard (high value)\",\n",
    "    \"Close to standard (low value)\",\n",
    "    \"Outlier (high value)\",\n",
    "    \"Outlier (low value)\",\n",
    "    \"Unknown\",\n",
    "]\n",
    "category_type = pd.CategoricalDtype(categories=categories, ordered=True)\n",
    "df[\"cpdReport\"] = df[\"cpdReport\"].astype(category_type)\n",
    "\n",
    "# sort by best scores\n",
    "df = df.sort_values(by=[\"busco\", \"cpdReport\"], ascending=[False, True])\n",
    "df = df.reset_index(drop=True)\n",
    "df.to_csv(\"tmp.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'buscoReport': {'complete': 2817,\n",
       "  'completeSingle': 2355,\n",
       "  'completeDuplicated': 462,\n",
       "  'fragmented': 43,\n",
       "  'missing': 74,\n",
       "  'total': 2934,\n",
       "  'lineageDb': 'arachnida_odb10',\n",
       "  'score': 97},\n",
       " 'cpdReport': {'proteomeCount': 0,\n",
       "  'stdCdss': 0.0,\n",
       "  'averageCdss': 0,\n",
       "  'confidence': 0,\n",
       "  'status': 'Unknown'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteom_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      Outlier (high value)\n",
       "1                  Standard\n",
       "2                  Standard\n",
       "3                  Standard\n",
       "4                  Standard\n",
       "               ...         \n",
       "314                 Unknown\n",
       "315                 Unknown\n",
       "316                Standard\n",
       "317    Outlier (high value)\n",
       "318                 Unknown\n",
       "Name: cpdReport, Length: 319, dtype: category\n",
       "Categories (6, object): ['Standard' < 'Close to standard (high value)' < 'Close to standard (low value)' < 'Outlier (high value)' < 'Outlier (low value)' < 'Unknown']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cpdReport\"]  # .value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'get'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb Cell 2\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m top_x_proteomes\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m \u001b[39m# Call the function and print the top X proteomes\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m top_busco_proteomes \u001b[39m=\u001b[39m find_top_busco_proteomes(proteom_json_path, x\u001b[39m=\u001b[39;49m\u001b[39m10\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m top_busco_proteomes\n",
      "\u001b[1;32m/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb Cell 2\u001b[0m line \u001b[0;36m9\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m proteome_busco_scores \u001b[39m=\u001b[39m []\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39mfor\u001b[39;00m entry \u001b[39min\u001b[39;00m proteomes_data:\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     proteome_id \u001b[39m=\u001b[39m entry\u001b[39m.\u001b[39;49mget(\u001b[39m'\u001b[39m\u001b[39mid\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     busco_score \u001b[39m=\u001b[39m entry\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mproteomeCompletenessReport\u001b[39m\u001b[39m'\u001b[39m, {})\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mbuscoReport\u001b[39m\u001b[39m'\u001b[39m, {})\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mscore\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m0\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/tobiassenoner/Documents/projects/phylogeny/src/prot_family/parse_busco.ipynb#W1sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m     proteome_busco_scores\u001b[39m.\u001b[39mappend((proteome_id, busco_score))\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'get'"
     ]
    }
   ],
   "source": [
    "# Now that we have verified the presence of the 'buscoReport' field and its structure, we can parse the JSON file.\n",
    "# We will load the proteome data and extract the proteome IDs along with their BUSCO 'complete' scores.\n",
    "\n",
    "# Define a function to find top X proteomes with the highest BUSCO score\n",
    "def find_top_busco_proteomes(proteomes_data, x=10):\n",
    "    # Extract proteome ID and BUSCO complete scores\n",
    "    proteome_busco_scores = []\n",
    "    for entry in proteomes_data:\n",
    "        proteome_id = entry.get(\"id\", \"\")\n",
    "        busco_score = (\n",
    "            entry.get(\"proteomeCompletenessReport\", {})\n",
    "            .get(\"buscoReport\", {})\n",
    "            .get(\"score\", 0)\n",
    "        )\n",
    "        proteome_busco_scores.append((proteome_id, busco_score))\n",
    "\n",
    "    # Sort by BUSCO score in descending order\n",
    "    proteome_busco_scores.sort(key=lambda entry: entry[1], reverse=True)\n",
    "\n",
    "    # Select the top X proteomes\n",
    "    top_x_proteomes = proteome_busco_scores[:x]\n",
    "\n",
    "    return top_x_proteomes\n",
    "\n",
    "\n",
    "# Call the function and print the top X proteomes\n",
    "top_busco_proteomes = find_top_busco_proteomes(proteom_json_path, x=10)\n",
    "top_busco_proteomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rostlab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
